src/shadepkg/game/physicsbody.nim

```diff
@@ -107,54 +110,37 @@ proc getBounds*(this: PhysicsBody): Rectangle =
     this.bounds = this.collisionShape.getBounds().getTranslatedInstance(this.getLocation())
   return this.bounds
 
-proc addCollisionListenerNow*(this: PhysicsBody, listener: CollisionListener, fireOnce: bool = false) =
-  if fireOnce:
-    var onceListener: CollisionListener
-    onceListener = proc(a, b: PhysicsBody, collisionResult: CollisionResult, gravityNormal: Vector) =
-      listener(a, b, collisionResult, gravityNormal)
-      this.removeCollisionListener(onceListener)
-    this.collisionListeners.add(onceListener)
-  else:
-    this.collisionListeners.add(listener)
+proc addCollisionListener(this: PhysicsBody, listener: CollisionListener) =
+  this.collisionListeners.add(listener)
 
-proc addCollisionListener*(this: PhysicsBody, listener: CollisionListener, fireOnce: bool = false) =
-  if tryAcquire(this.collisionListenersLock):
-    this.addCollisionListenerNow(listener, fireOnce)
-    this.collisionListenersLock.release()
-  else:
-    this.collisionListenersToAdd.addLast((listener, fireOnce))
-
-proc removeCollisionListenerNow*(this: PhysicsBody, listener: CollisionListener) =
-  var index = -1
-  for i, l in this.collisionListeners:
-    if l == listener:
-      index = i
-      break
-  
-  if index >= 0:
-    this.collisionListeners.delete(index)
+template buildCollisionListener*(thisBody: PhysicsBody, body: untyped) =
+  let listener: CollisionListener =
+    proc(
+      this {.inject.}, other {.inject.}: PhysicsBody,
+      collisionResult {.inject.}: CollisionResult,
+      gravityNormal {.inject.}: Vector
+    ): bool =
+      body
+
+  thisBody.addCollisionListener(listener)
 
 proc removeCollisionListener*(this: PhysicsBody, listener: CollisionListener) =
-  if tryAcquire(this.collisionListenersLock):
-    this.removeCollisionListenerNow(listener)
-    this.collisionListenersLock.release()
-  else:
-    this.collisionListenersToRemove.addLast(listener)
+  this.collisionListeners.remove(listener)
 
 proc notifyCollisionListeners*(
   this, other: PhysicsBody,
   collisionResult: CollisionResult,
   gravityNormal: Vector
 ) =
-  withLock(this.collisionListenersLock):
-    for listener in this.collisionListeners:
-      listener(this, other, collisionResult, gravityNormal)
+  for listener in this.collisionListeners:
+    if listener(this, other, collisionResult, gravityNormal):
+      this.collisionListeners.remove(listener)
 
 proc wallAndGroundSetter(
   this, other: PhysicsBody,
   collisionResult: CollisionResult,
   gravityNormal: Vector
-) =
+): bool =
```

Comment:
Why was a return type added but no return statement? Does Nim have implicit falsy return?

Reply:
Nim has an implicit variable named `result` that matches the return type, and it is the default value for that type (anything 0'd out, so false for booleans, nil for ref objects)

So if the user does not explicitly return a value, `false` is returned, so the listener is not removed after being triggered